### PR TITLE
ALSA: hda: avoid write to STATESTS if controller is in reset

### PR DESCRIPTION
**Pull request by kv2019i**
The HDA specification states that controller must be taken out
of reset before writing to registers other than GCTL.CRST (1.0a
spec, 3.3.7).

The snd_hdac_bus_reset_link() does an unconditional write to
STATESTS. Newer Intel HDA controllers will emit an error if
this is done when the controller is in reset. This has no impact
to HDA operation, but error may be logged e.g. via PCIe subsystem.

Fix the code to align with the spec and only clear STATESTS,
when controller is not in reset.

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: sound/hda/hdac_controller.c [hunk 1]
```diff
@@ -421,8 +421,9 @@ int snd_hdac_bus_reset_link(struct hdac_bus *bus, bool full_reset)
 	if (!full_reset)
 		goto skip_reset;
 
-	/* clear STATESTS */
-	snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
+	/* clear STATESTS if not in reset */
+	if (snd_hdac_chip_readb(bus, GCTL) & AZX_GCTL_RESET)
+		snd_hdac_chip_writew(bus, STATESTS, STATESTS_INT_MASK);
```
Comment (plbossart) [inline sound/hda/hdac_controller.c:426]: the description of the change is a bit odd. it's not clear if 

a) the driver shall skip the write to STATETS if not in reset
b) the driver shall wait until the controller is out of reset before writing to STATETS

Comment (ranj063) [reply sound/hda/hdac_controller.c:426]: apart from the above comment, Im also curious to know if the error is actually something we've seen ie we're clearing the STATETS when the the controller is already in reset? And if so, can we skip the call to reset the controller in the subsequent line?

Comment (kv2019i) [reply sound/hda/hdac_controller.c:426]: @plbossart @ranj063 Don't ask me, this code is again as ancient as computers themselves. This function is called in many cases and sometimes controller is not in reset, and for correct operation, STATESTS should be cleared. But this is also called in cases where controller is just powered on and is in reset. Based on HDA spec, in the latter case the write will go to /dev/null.

Now I have no idea if there are non-Intel controllers where the STATESTS write does not go to /dev/null, but as per HDA spec, this should not happen. But of course very tricky to test.

I'll try to work on the commit message.

**Comment by kv2019i** [comment]
I dug into the history of this STATESTS write (that I'm now proposing to sometimes skip), wrote a short novel about it and sent upstream:
https://mailman.alsa-project.org/pipermail/alsa-devel/2021-October/191065.html

Let's see how this goes. Closing here.

**Pull request closed without merging**